### DISCUSSION TITLE
How to Upgrade?

### DISCUSSION THREAD

**Discussion by fung** [digitalocean/Kubernetes-Starter-Kit-Developers#79] [Q&A] [answered]
Hi,

The `ambassador` has newer version (others as well). May I know what is the recommended approach to upgrade?

Thanks a lot!

**Comment by v-ctiutiu** [comment, marked as answer]
Hi @fung 

The recommended way is via `Helm`. Please perform a `backup` of your `CRDs` first, in case something goes wrong.

For example, if you want to upgrade `Ambassador` to the latest version please use the following:

1. Fetch the latest Helm chart version:

```shell
helm repo update
```

2. List new versions available in the charts repository:

```shell
helm search repo datawire/ambassador
```

The output looks similar to (notice that `6.9.3` seems to be the latest chart version, which maps to application version `1.14.2`):

```text
NAME                            CHART VERSION   APP VERSION     DESCRIPTION                         
datawire/ambassador             6.9.3           1.14.2          A Helm chart for Datawire Ambassador
datawire/ambassador-operator    0.3.0           v1.3.0          A Helm chart for Kubernetes
```

3. Upgrade CRDs (starting with Helm 3, it doesn't happen automatically)

```shell
kubectl apply -f https://www.getambassador.io/yaml/aes-crds.yaml
```

4. Perform an upgrade using the latest chart version (`6.9.3` as per the above listing):

```shell
helm upgrade ambassador datawire/ambassador --version 6.9.3 --namespace ambassador -f <YOUR_VALUES_FILE>.yaml
```

**Hint:**

If you don't rely on a specific major version, and just want to upgrade to the latest no matter what, then all you have to do is to omit the `--version` argument:

```shell
helm upgrade ambassador datawire/ambassador --namespace ambassador -f <YOUR_VALUES_FILE>.yaml
```

Please bear in mind that the `Starter Kit` is using the `https://www.getambassador.io` repository for `Helm` charts, which may not contain the latest version for `Ambassador` (which seems to be `2.0.x` now). This is so because at that time of writing we had `1.13.10` as the latest `stable` version `available`.

Also, please note that the chart version is not the same with application version. In the end, you're interested in the application version which is the `APP VERSION` column value from the `helm search repo datawire/ambassador` command output.

The same set of steps can be used for the rest of the `Starter Kit` components if they were installed via `command line` using `Helm`.

If you use `Flux CD`, then all you need to do is to update the helm release manifest file from your `Git` repository. The chart version is available in the `spec.chart.spec.version` field:

```yaml
spec:
  chart:
    spec:
      chart: ambassador
      sourceRef:
        kind: HelmRepository
        name: ambassador
      version: 6.7.13
```

Looking at the above snippet for `Ambassador`, you need to modify the `version` field and commit the changes to your `Git` repository synced by `Flux`.

Now, if you want to `upgrade` to a `major release`, then it's best to follow the official documentation because there may be some extra steps to take care of. Usually it's done via `Helm` using the `helm upgrade` command explained earlier. For `Ambassador` you can follow the official [migration guide](https://www.getambassador.io/docs/edge-stack/latest/topics/install/migrate-to-version-2), if you want to go from `1.x.x` to `2.0.x`. 

I also noticed that `Ambassador` has a new repository for Helm charts as well: `https://app.getambassador.io`, so you need to point to that one if you want to go with version `2.0.x`.

In the future, we will add an `upgrade section` for each `component` of the `Starter Kit`.

Hope it helps.

**Reply by fung** [reply]
Hi @v-ctiutiu ,

Many thanks for the detailed explanation! I learned a lot from this repo and other resources on DO as well.

Thanks DO teams!
